src/script/lua_api/l_particles_local.cpp

```diff
@@ -0,0 +1,244 @@
+	v3f pos, vel, acc;
+	pos = vel = acc = v3f(0, 0, 0);
+
+	float expirationtime, size;
+	expirationtime = size = 1;
+
+	bool collisiondetection, vertical, collision_removal;
+	collisiondetection = vertical = collision_removal = false;
+	struct TileAnimationParams animation;
+	animation.type = TAT_NONE;
+
+	std::string texture = "";
+
+	u8 glow = 0;
+
+	lua_getfield(L, 1, "pos");
+	pos = lua_istable(L, -1) ? check_v3f(L, -1) : v3f();
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "velocity");
+	vel = lua_istable(L, -1) ? check_v3f(L, -1) : vel;
+	lua_pop(L, 1);
+
+	lua_getfield(L, 1, "acceleration");
+	acc = lua_istable(L, -1) ? check_v3f(L, -1) : acc;
+	lua_pop(L, 1);
+
+	expirationtime = getfloatfield_default(L, 1, "expirationtime", 1);
+	size = getfloatfield_default(L, 1, "size", 1);
+	collisiondetection =
+		getboolfield_default(L, 1, "collisiondetection", collisiondetection);
+	collision_removal =
+		getboolfield_default(L, 1, "collision_removal", collision_removal);
+	vertical = getboolfield_default(L, 1, "vertical", vertical);
+
+	lua_getfield(L, 1, "animation");
+	animation = read_animation_definition(L, -1);
+	lua_pop(L, 1);
+
+	texture = getstringfield_default(L, 1, "texture", "");
+
+	glow = getintfield_default(L, 1, "glow", 0);
+
+	ClientEvent event;
+	event.type                              = CE_SPAWN_PARTICLE;
+	event.spawn_particle.pos                = new v3f (pos);
+	event.spawn_particle.vel                = new v3f (vel);
+	event.spawn_particle.acc                = new v3f (acc);
+	event.spawn_particle.expirationtime     = expirationtime;
+	event.spawn_particle.size               = size;
+	event.spawn_particle.collisiondetection = collisiondetection;
+	event.spawn_particle.collision_removal  = collision_removal;
+	event.spawn_particle.vertical           = vertical;
+	event.spawn_particle.texture            = new std::string(texture);
+	event.spawn_particle.animation          = animation;
+	event.spawn_particle.glow               = glow;
+	getClient(L)->pushToEventQueue(event);
+
+	return 0;
+}
+
+// add_particlespawner({amount=, time=,
+//				minpos=, maxpos=,
+//				minvel=, maxvel=,
+//				minacc=, maxacc=,
+//				minexptime=, maxexptime=,
+//				minsize=, maxsize=,
+//				collisiondetection=,
+//				collision_removal=,
+//				vertical=,
+//				texture=})
+// minpos/maxpos/minvel/maxvel/minacc/maxacc = {x=num, y=num, z=num}
+// minexptime/maxexptime = num (seconds)
+// minsize/maxsize = num
+// collisiondetection = bool
+// collision_removal = bool
+// vertical = bool
+// texture = e.g."default_wood.png"
+// animation = TileAnimation definition
+// glow = num
+int ModApiParticlesLocal::l_add_particlespawner(lua_State *L)
+{
+	luaL_checktype(L, 1, LUA_TTABLE);
+	// Get parameters
+	u16 amount = 1;
+	v3f minpos, maxpos, minvel, maxvel, minacc, maxacc;
+	minpos = maxpos = minvel = maxvel = minacc = maxacc = v3f(0, 0, 0);
+	float time, minexptime, maxexptime, minsize, maxsize;
+	time = minexptime = maxexptime = minsize = maxsize = 1;
```

Comment:
IMO, this would look better if the default values were set in the conditional expressions below.